### PR TITLE
[DataGridPro] Fix header filtering with `boolean` column type

### PR DESCRIPTION
**Pull request by k-rajat19**
- [x] I have followed (at least) the [PR section of the contributing guide](https://github.com/mui/mui-x/blob/HEAD/CONTRIBUTING.md#sending-a-pull-request).

Closes https://github.com/mui/mui-x/issues/15524

Related to https://github.com/mui/mui-x/pull/15252
Preview - https://deploy-preview-15528--material-ui-x.netlify.app/x/react-data-grid/filtering/header-filters/

### CODE REVIEW

File: packages/x-data-grid-pro/src/components/headerFiltering/GridHeaderFilterCell.tsx [hunk 1]
```diff
@@ -167,7 +167,7 @@ const GridHeaderFilterCell = React.forwardRef<HTMLDivElement, GridHeaderFilterCe
 
     const applyFilterChanges = React.useCallback(
       (updatedItem: GridFilterItem) => {
-        if (item.value && !updatedItem.value) {
+        if (item.value && updatedItem.value === undefined) {
```
Comment (michelengelen) [inline packages/x-data-grid-pro/src/components/headerFiltering/GridHeaderFilterCell.tsx:170]:
```suggestion
        if (item.value && updatedItem.value == undefined) {
```

to allow checking for `null` as well ... just to be safe!

File: packages/x-data-grid-pro/src/components/headerFiltering/GridHeaderFilterCell.tsx [hunk 2]
```diff
@@ -314,7 +314,7 @@ const GridHeaderFilterCell = React.forwardRef<HTMLDivElement, GridHeaderFilterCe
 
     const isNoInputOperator = currentOperator.requiresFilterValue === false;
 
-    const isApplied = Boolean(item?.value) || isNoInputOperator;
+    const isApplied = item?.value !== undefined || isNoInputOperator;
```
Comment (michelengelen) [inline packages/x-data-grid-pro/src/components/headerFiltering/GridHeaderFilterCell.tsx:317]: same here
```suggestion
    const isApplied = item?.value != undefined || isNoInputOperator;
```

**Review by michelengelen — COMMENTED** [review]
just 2 nitpicks

File: packages/x-data-grid-pro/src/components/headerFiltering/GridHeaderFilterCell.tsx (hunk 1 above)
Comment (k-rajat19) [reply packages/x-data-grid-pro/src/components/headerFiltering/GridHeaderFilterCell.tsx:170]: thanks for the suggestion
`null`  values already handled here https://github.com/mui/mui-x/blob/8ee87b5689b772634f2ced89e83493a095b9a681/packages/x-data-grid/src/components/panel/filterPanel/GridFilterInputBoolean.tsx#L26
I have also added test for it 2d3521d

**Comment by michelengelen** [comment]
all right ... I'll let @arminmeh as the main reviewer approve this! 👍🏼

**Comment by arminmeh** [comment]
@k-rajat19 should this be merged?

We usually wait for the contributors to merge their own PRs (in case there is still something that they had in mind)

**Comment by k-rajat19** [comment]
All good from my side, you can merge it
I don't have permissions to merge PRs.

**Pull request merged**